### PR TITLE
Table Batch Updates

### LINKED ISSUE #1481: Azurite delete_entity is unable to handle keys containing an apostrophe (unlike Azure Storage account in the cloud)
### Which service(blob, file, queue, table) does this issue concern?

table

### Which version of the Azurite was used?

3.17.1

### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)

npm

### What's the Node.js version?

v16.14.0

### What problem was encountered?

The Azurite emulator is unable to handle apostrophes in keys for the delete_entity method. This is unlike Azure Storage account in the cloud which can handle them fine.

### Steps to reproduce the issue?

This Python code demonstrates the issue (this is using Python v3.9.9 and azure-data-tables v12.3.0):

  from azure.data.tables import TableClient, TableServiceClient
  
  CONNECTION_STRING = (
      "DefaultEndpointsProtocol=https;AccountName=devstoreaccount1;"
      + "AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;"
      + "BlobEndpoint=https://127.0.0.1:10000/devstoreaccount1;QueueEndpoint=https://127.0.0.1:10001/devstoreaccount1;"
      + "TableEndpoint=http://127.0.0.1:10002/devstoreaccount1;"
  )
  TABLE_NAME = "TestTable"
  
  PARTITION_KEY = "David"
  ROW_KEY = "O'Reilly"  # This is the value that causes an issue
  
[debug.log](https://github.com/Azure/Azurite/files/8644796/debug.log)
[debug.log](https://github.com/Azure/Azurite/files/8644797/debug.log)

  ENTITY = {"PartitionKey": PARTITION_KEY, "RowKey": ROW_KEY, "address": "1 Made Up Street"}
  
  
  def count_entities(table_client: TableClient) -> None:
      entities = table_client.list_entities()
      entity_count = 0
      for _ in entities:
          entity_count += 1
      print(f"Entity Count: {entity_count}")
  
  
  service = TableServiceClient.from_connection_string(conn_str=CONNECTION_STRING)
  table_client = service.get_table_client(TABLE_NAME)
  table_client.create_entity(ENTITY)
  count_entities(table_client)
  table_client.delete_entity(partition_key=PARTITION_KEY, row_key=ROW_KEY)
  print("** Attempted Deletion **")
  count_entities(table_client)

The code above when run against Azurite produces the following output:

  Entity Count: 1
  ** Attempted Deletion **
  Entity Count: 1

But when the connection string in code above is changed to an Azure Storage account in the cloud, it produces a different output:

  Entity Count: 1
  ** Attempted Deletion **
  Entity Count: 0

[debug.log](https://github.com/Azure/Azurite/files/8644799/debug.log)

### Have you found a mitigation/solution?

No. I am also unable to delete the row using Microsoft Azure Storage Explorer. The only workaround I've found is to delete and recreate the table.

### LINKED ISSUE #1754: Failed table transaction with multiple entities with same RowKey returns HTTP 409(EntityAlreadyExists) instead of HTTP 400(InvalidDuplicateRow)
### Which service(blob, file, queue, table) does this issue concern?
table

### Which version of the Azurite was used?
3.20.1

### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)
npm

### What's the Node.js version?
v12.18.4

### What problem was encountered?
When submitting a batch transaction which adding more than one elements which same rowKey got inappropriate error.

`1:The specified entity already exists.
RequestId:8a67323b-cde6-41d5-89eb-79ba4ad1b7dc
Time:2022-11-16T14:31:52.155Z
 The index of the entity that caused the error can be found in FailedTransactionActionIndex.
Status: 409 (Conflict)
ErrorCode: EntityAlreadyExists`

The proper error should be 400 (InvalidDuplicateRow) like this:
`1:The batch request contains multiple changes with same row key. An entity can appear only once in a batch request.
RequestId:37cdf7a6-5873-4efb-8b94-be36221a6b7a
Time:2022-11-16T14:33:57.5727871Z
 The index of the entity that caused the error can be found in FailedTransactionActionIndex.
Status: 400 (Bad Request)
ErrorCode: InvalidDuplicateRow`

### Steps to reproduce the issue?
```csharp
var tableServiceClient = new TableServiceClient("UseDevelopmentStorage=true;");
var tableClient = tableServiceClient.GetTableClient("sampleTable");
await tableClient.CreateIfNotExistsAsync();

try
{
    await tableClient.SubmitTransactionAsync(new List<TableTransactionAction>
    {
         new(TableTransactionActionType.Add, new TableEntity("samePartition", "sameRowKey")),
         new(TableTransactionActionType.Add, new TableEntity("samePartition", "sameRowKey"))
    });
}
catch (RequestFailedException e)
{
      Console.WriteLine(e.ToString());
}
```

If possible, please provide the debug log using the -d parameter, replacing \<pathtodebuglog\> with an appropriate path for your OS, or review the instructions for docker containers:

```2022-11-16T15:13:47.538Z 	 info: Azurite Table service is starting on 127.0.0.1:10002
2022-11-16T15:13:47.539Z 	 info: AccountDataStore:init() Refresh accounts from environment variable AZURITE_ACCOUNTS with value undefined
2022-11-16T15:13:47.539Z 	 info: AccountDataStore:init() Fallback to default emulator account devstoreaccount1.
2022-11-16T15:13:47.567Z 	 info: Azurite Table service successfully listens on http://127.0.0.1:10002
2022-11-16T15:13:52.552Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: TableStorageContextMiddleware: RequestMethod=POST RequestURL=http://127.0.0.1/devstoreaccount1/Tables?$format=application%2Fjson%3Bodata%3Dminimalmetadata RequestHeaders:{"host":"127.0.0.1:10002","x-ms-version":"2020-12-06","dataserviceversion":"3.0","prefer":"return-no-content","accept":"application/json;odata=minimalmetadata","x-ms-client-request-id":"a825bcd7-9063-4234-92eb-00e6546f0130","x-ms-return-client-request-id":"true","user-agent":"azsdk-net-Data.Tables/12.7.1 (.NET 6.0.11; Microsoft Windows 10.0.22000)","x-ms-date":"Wed, 16 Nov 2022 15:13:52 GMT","authorization":"SharedKeyLite devstoreaccount1:J3PgAFlOxa7iVl6rarqpYP6C8KraJtpOZ9Ng9i0Ti94=","content-type":"application/json;odata=nometadata","content-length":"27"} ClientIP=127.0.0.1 Protocol=http HTTPVersion=1.1
2022-11-16T15:13:52.553Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 debug: tableStorageContextMiddleware: Dispatch pattern string: /Tables
2022-11-16T15:13:52.553Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: tableStorageContextMiddleware: Account=devstoreaccount1 tableName=undefined
2022-11-16T15:13:52.553Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 verbose: DispatchMiddleware: Dispatching request...
2022-11-16T15:13:52.555Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: DispatchMiddleware: Operation=Table_Create
2022-11-16T15:13:52.555Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 verbose: AuthenticationMiddlewareFactory:createAuthenticationMiddleware() Validating authentications.
2022-11-16T15:13:52.555Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: TableSharedKeyLiteAuthenticator:validate() Start validation against account shared key authentication.
2022-11-16T15:13:52.556Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: TableSharedKeyLiteAuthenticator:validate() [STRING TO SIGN]:"Wed, 16 Nov 2022 15:13:52 GMT\n/devstoreaccount1/devstoreaccount1/Tables"
2022-11-16T15:13:52.557Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: TableSharedKeyLiteAuthenticator:validate() Calculated authentication header based on key1: SharedKeyLite devstoreaccount1:J3PgAFlOxa7iVl6rarqpYP6C8KraJtpOZ9Ng9i0Ti94=
2022-11-16T15:13:52.557Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: TableSharedKeyLiteAuthenticator:validate() Signature 1 matched.
2022-11-16T15:13:52.557Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 verbose: DeserializerMiddleware: Start deserializing...
2022-11-16T15:13:52.559Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 debug: deserialize(): Raw request body string is (removed all empty characters) {"TableName":"sampleTable"}
2022-11-16T15:13:52.560Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: HandlerMiddleware: DeserializedParameters={"options":{"queryOptions":{"format":"application/json;odata=minimalmetadata"},"requestId":"a825bcd7-9063-4234-92eb-00e6546f0130","dataServiceVersion":"3.0","responsePreference":"return-no-content"},"version":"2020-12-06","tableProperties":{"tableName":"sampleTable"},"body":"ReadableStream"}
2022-11-16T15:13:52.563Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: Received a MiddlewareError, fill error information to HTTP response
2022-11-16T15:13:52.564Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: ErrorName=StorageError ErrorMessage="The table specified already exists."  ErrorHTTPStatusCode=409 ErrorHTTPStatusMessage=undefined ErrorHTTPHeaders={"x-ms-error-code":"TableAlreadyExists","x-ms-request-id":"50422e6b-4c6a-4fd4-adef-d6f6ac3545e1","x-ms-version":"2021-10-04"} ErrorHTTPBody="{\"odata.error\":{\"code\":\"TableAlreadyExists\",\"message\":{\"lang\":\"en-US\",\"value\":\"The table specified already exists.\\nRequestId:50422e6b-4c6a-4fd4-adef-d6f6ac3545e1\\nTime:2022-11-16T15:13:52.562Z\"}}}" ErrorStack="\"StorageError: The table specified already exists.\\n    at Function.getTableAlreadyExists (C:\\\\azurite\\\\node_modules\\\\azurite\\\\dist\\\\src\\\\table\\\\errors\\\\StorageErrorFactory.js:23:16)\\n    at LokiTableMetadataStore.createTable (C:\\\\azurite\\\\node_modules\\\\azurite\\\\dist\\\\src\\\\table\\\\persistence\\\\LokiTableMetadataStore.js:114:49)\\n    at TableHandler.create (C:\\\\azurite\\\\node_modules\\\\azurite\\\\dist\\\\src\\\\table\\\\handlers\\\\TableHandler.js:40:34)\\n    at C:\\\\azurite\\\\node_modules\\\\azurite\\\\dist\\\\src\\\\table\\\\generated\\\\middleware\\\\HandlerMiddlewareFactory.js:58:18\\n    at C:\\\\azurite\\\\node_modules\\\\azurite\\\\dist\\\\src\\\\table\\\\generated\\\\ExpressMiddlewareFactory.js:77:63\\n    at Layer.handle [as handle_request] (C:\\\\azurite\\\\node_modules\\\\azurite\\\\node_modules\\\\express\\\\lib\\\\router\\\\layer.js:95:5)\\n    at trim_prefix (C:\\\\azurite\\\\node_modules\\\\azurite\\\\node_modules\\\\express\\\\lib\\\\router\\\\index.js:328:13)\\n    at C:\\\\azurite\\\\node_modules\\\\azurite\\\\node_modules\\\\express\\\\lib\\\\router\\\\index.js:286:9\\n    at Function.process_params (C:\\\\azurite\\\\node_modules\\\\azurite\\\\node_modules\\\\express\\\\lib\\\\router\\\\index.js:346:12)\\n    at next (C:\\\\azurite\\\\node_modules\\\\azurite\\\\node_modules\\\\express\\\\lib\\\\router\\\\index.js:280:10)\""
2022-11-16T15:13:52.564Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: Set HTTP code: 409
2022-11-16T15:13:52.564Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: Set HTTP Header: x-ms-error-code=TableAlreadyExists
2022-11-16T15:13:52.564Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: Set HTTP Header: x-ms-request-id=50422e6b-4c6a-4fd4-adef-d6f6ac3545e1
2022-11-16T15:13:52.564Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: Set HTTP Header: x-ms-version=2021-10-04
2022-11-16T15:13:52.564Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: Set content type: application/json;odata=minimalmetadata;streaming=true;charset=utf-8
2022-11-16T15:13:52.564Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 error: ErrorMiddleware: Set HTTP body: "{\"odata.error\":{\"code\":\"TableAlreadyExists\",\"message\":{\"lang\":\"en-US\",\"value\":\"The table specified already exists.\\nRequestId:50422e6b-4c6a-4fd4-adef-d6f6ac3545e1\\nTime:2022-11-16T15:13:52.562Z\"}}}"
2022-11-16T15:13:52.565Z 50422e6b-4c6a-4fd4-adef-d6f6ac3545e1 info: EndMiddleware: End response. TotalTimeInMS=13 StatusCode=409 StatusMessage=Conflict Headers={"server":"Azurite-Table/3.20.1","x-ms-error-code":"TableAlreadyExists","x-ms-request-id":"50422e6b-4c6a-4fd4-adef-d6f6ac3545e1","x-ms-version":"2021-10-04","content-type":"application/json;odata=minimalmetadata;streaming=true;charset=utf-8"}
2022-11-16T15:13:52.681Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: TableStorageContextMiddleware: RequestMethod=POST RequestURL=http://127.0.0.1/devstoreaccount1/$batch RequestHeaders:{"host":"127.0.0.1:10002","x-ms-version":"2020-12-06","dataserviceversion":"3.0","accept":"application/json","x-ms-client-request-id":"5bc9c666-f106-4d3e-91b5-167ed73b61f8","x-ms-return-client-request-id":"true","user-agent":"azsdk-net-Data.Tables/12.7.1 (.NET 6.0.11; Microsoft Windows 10.0.22000)","x-ms-date":"Wed, 16 Nov 2022 15:13:52 GMT","authorization":"SharedKeyLite devstoreaccount1:W6l0867Tt0iztUICd3Q9Bcd9mVXCBjGnEgW/K6Jrips=","content-type":"multipart/mixed; boundary=batch_03995ca5-001f-440d-82f8-11b40121dd5e","content-length":"1216"} ClientIP=127.0.0.1 Protocol=http HTTPVersion=1.1
2022-11-16T15:13:52.681Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 debug: tableStorageContextMiddleware: Dispatch pattern string: /$batch
2022-11-16T15:13:52.681Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: tableStorageContextMiddleware: Account=devstoreaccount1 tableName=$batch
2022-11-16T15:13:52.681Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 verbose: DispatchMiddleware: Dispatching request...
2022-11-16T15:13:52.682Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: DispatchMiddleware: Operation=Table_Batch
2022-11-16T15:13:52.682Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 verbose: AuthenticationMiddlewareFactory:createAuthenticationMiddleware() Validating authentications.
2022-11-16T15:13:52.682Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: TableSharedKeyLiteAuthenticator:validate() Start validation against account shared key authentication.
2022-11-16T15:13:52.682Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: TableSharedKeyLiteAuthenticator:validate() [STRING TO SIGN]:"Wed, 16 Nov 2022 15:13:52 GMT\n/devstoreaccount1/devstoreaccount1/$batch"
2022-11-16T15:13:52.682Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: TableSharedKeyLiteAuthenticator:validate() Calculated authentication header based on key1: SharedKeyLite devstoreaccount1:W6l0867Tt0iztUICd3Q9Bcd9mVXCBjGnEgW/K6Jrips=
2022-11-16T15:13:52.682Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: TableSharedKeyLiteAuthenticator:validate() Signature 1 matched.
2022-11-16T15:13:52.682Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 verbose: DeserializerMiddleware: Start deserializing...
2022-11-16T15:13:52.683Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: HandlerMiddleware: DeserializedParameters={"version":"2020-12-06","options":{"requestId":"5bc9c666-f106-4d3e-91b5-167ed73b61f8","dataServiceVersion":"3.0"},"multipartContentType":"multipart/mixed; boundary=batch_03995ca5-001f-440d-82f8-11b40121dd5e","contentLength":1216,"body":"ReadableStream"}
2022-11-16T15:13:52.684Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 debug: TableHandler:batch() Raw request string is "--batch_03995ca5-001f-440d-82f8-11b40121dd5e\r\nContent-Type: multipart/mixed; boundary=changeset_6ec27c11-728f-4b70-b763-a8cbbf761fe6\r\n\r\n--changeset_6ec27c11-728f-4b70-b763-a8cbbf761fe6\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nPOST http://127.0.0.1:10002/devstoreaccount1/sampleTable?$format=application%2Fjson%3Bodata%3Dminimalmetadata HTTP/1.1\r\nHost: 127.0.0.1\r\nx-ms-version: 2020-12-06\r\nDataServiceVersion: 3.0\r\nPrefer: return-no-content\r\nAccept: application/json;odata=minimalmetadata\r\nContent-Type: application/json;odata=nometadata\r\n\r\n{\"PartitionKey\":\"samePartition\",\"RowKey\":\"sameRowKey\"}\r\n--changeset_6ec27c11-728f-4b70-b763-a8cbbf761fe6\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nPOST http://127.0.0.1:10002/devstoreaccount1/sampleTable?$format=application%2Fjson%3Bodata%3Dminimalmetadata HTTP/1.1\r\nHost: 127.0.0.1\r\nx-ms-version: 2020-12-06\r\nDataServiceVersion: 3.0\r\nPrefer: return-no-content\r\nAccept: application/json;odata=minimalmetadata\r\nContent-Type: application/json;odata=nometadata\r\n\r\n{\"PartitionKey\":\"samePartition\",\"RowKey\":\"sameRowKey\"}\r\n--changeset_6ec27c11-728f-4b70-b763-a8cbbf761fe6--\r\n\r\n--batch_03995ca5-001f-440d-82f8-11b40121dd5e--\r\n"
2022-11-16T15:13:52.692Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 debug: TableHandler:batch() Raw response string is "--batchresponse_03995ca5-001f-440d-82f8-11b40121dd5e\r\nContent-Type: multipart/mixed; boundary=changesetresponse_6ec27c11-728f-4b70-b763-a8cbbf761fe6\r\n\r\n--changesetresponse_6ec27c11-728f-4b70-b763-a8cbbf761fe6\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nHTTP/1.1 409 Conflict\r\nContent-ID: 2\r\nDataServiceVersion: 3.0;\r\nContent-Type: application/json;odata=minimalmetadata;charset=utf-8\r\n\r\n{\"odata.error\":{\"code\":\"EntityAlreadyExists\",\"message\":{\"lang\":\"en-US\",\"value\":\"1:The specified entity already exists.\\nRequestId:b9a8bad9-e3cc-40b3-91d5-de5602d32871\\nTime:2022-11-16T15:13:52.690Z\"}}}\r\n--changesetresponse_6ec27c11-728f-4b70-b763-a8cbbf761fe6--\r\n--batchresponse_03995ca5-001f-440d-82f8-11b40121dd5e--\r\n"
2022-11-16T15:13:52.692Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 verbose: SerializerMiddleware: Start serializing...
2022-11-16T15:13:52.692Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: Serializer: Start returning stream body.
2022-11-16T15:13:52.695Z b9a8bad9-e3cc-40b3-91d5-de5602d32871 info: EndMiddleware: End response. TotalTimeInMS=14 StatusCode=202 StatusMessage=Accepted Headers={"server":"Azurite-Table/3.20.1","content-type":"multipart/mixed; boundary=batchresponse_03995ca5-001f-440d-82f8-11b40121dd5e","x-ms-request-id":"b9a8bad9-e3cc-40b3-91d5-de5602d32871","x-ms-version":"2021-10-04","date":"Wed, 16 Nov 2022 15:13:52 GMT"}
```

Please be sure to remove any PII or sensitive information before sharing!  
The debug log will log raw request headers and bodies, so that we can replay these against Azurite using REST and create tests to validate resolution.

### PR DESCRIPTION
**Pull request by edwin-huber**
As discussed, refactoring Table Batch.  
Added test and fixed error message #1754  
Fixes #1754  
Added table API test to cover #1726
Added query interpreter unit test to cover #1599 
Fixes #1481 when trying to delete rows with apostrophe.  
Fixes https://github.com/Azure/Azurite/issues/1958
Added test case for https://github.com/Azure/Azurite/issues/1958 in C# project as no repro via Javascript Data-Tables SDK.

### CODE REVIEW

**Comment by blueww** [comment]
@edwin-huber 
Thanks for the fixing!
Not sure if these issues related with table batch are also fixed? https://github.com/Azure/Azurite/labels/table-batch

**Comment by edwin-huber** [comment]
Thanks Wei Wei, I'll check all of them, the goal is to do some housekeeping and clear up as many batch related issues as possible and know which need significant work to fix.

File: src/table/batch/TableBatchOrchestrator.ts [hunk 1]
```diff
@@ -597,6 +658,8 @@ export default class TableBatchOrchestrator {
       // keys can have more complex values which are URI encoded if they come from the URL
       // we decode above.
       partitionKey = partKeyMatch[0];
+      // Url should use double ticks and we need to remove them
+      partitionKey = this.replaceDoubleTicks(partitionKey);
```
Comment (blueww) [inline src/table/batch/TableBatchOrchestrator.ts:662]: When a partition key really contains 2 continually ticks like `partition''key`, will this work?

File: tests/table/apis/table.entity.azure.data-tables.test.ts [hunk 2]
```diff
@@ -1280,4 +1295,232 @@ describe("table Entity APIs test - using Azure/data-tables", () => {
...
+    const tableClient = createAzureDataTablesClient(
+      testLocalAzuriteInstance,
+      getUniqueName("apostrophe")
+    );
+    await tableClient.createTable();
+    const percentPartition = "percentRowBatch";
+    const testEntity: TableTestEntity =
+      entityFactory.createBasicEntityForTest(percentPartition);
+
+    testEntity.rowKey = testEntity.rowKey.replace("row", "O'");
```
Comment (blueww) [inline tests/table/apis/table.entity.azure.data-tables.test.ts:1310]: We might can also check a row key include 2 ' like `row''key`

File: tests/table/apis/table.entity.rest.test.ts [hunk 3]
```diff
@@ -313,7 +313,7 @@ describe("table Entity APIs REST tests", () => {
     assert.strictEqual(weMerged, 1);
   });
 
-  it("Should not be able to use query enties in a batch request, @loki", async () => {
+  it("Should be able to use query enties in a batch request, @loki", async () => {
```
Comment (blueww) [inline tests/table/apis/table.entity.rest.test.ts:316]: The title looks not fully reflect what the case test. Might should be something similar to: Should be able to use query entities in a batch request on no exist entity

File: tests/table/dotnet/AzuriteTableTest/Program.cs [hunk 4]
```diff
@@ -14,10 +14,10 @@ class Program
         static async Task Main(string[] args)
         {
             // delete batch with cosmos
-            await TestForIssue1439.RunTest();
```
Comment (blueww) [inline tests/table/dotnet/AzuriteTableTest/Program.cs:17]: Why remove running test for issue 1439?

File: tests/table/dotnet/AzuriteTableTest/Program.cs [hunk 5]
```diff
@@ -14,10 +14,10 @@ class Program
...
+            await TestForIssue1958.RunTest();
 
-            // delete batch with data Tables
-            await TestForDataTablesDeleteBatch.RunTest();
+            // // delete batch with data Tables
+            // await TestForDataTablesDeleteBatch.RunTest();
```
Comment (blueww) [inline tests/table/dotnet/AzuriteTableTest/Program.cs:20]: Why skip running TestForDataTablesDeleteBatch()?

**Comment by blueww** [comment]
@edwin-huber 

Xiaoning, Emma and I have reviewed this PR.
The PR generally looks good to us.
We have raised several comments, would you please help to review them?

We can merge this PR after all these comments are resolved.

File: src/table/batch/TableBatchOrchestrator.ts (hunk 1 above)
Comment (edwin-huber) [reply src/table/batch/TableBatchOrchestrator.ts:662]: I added an additional test file to cover these cases...  
Based on string escaping logic, this should be correct... but agree we need more test coverage to ensure all cases.
What we can be sure of, is that SDK is using double tick escaping in the request, so we need to find a way to check all cases, or determine which type of escaping is used in the request to react accordingly.

File: tests/table/apis/table.entity.azure.data-tables.test.ts (hunk 2 above)
Comment (edwin-huber) [reply tests/table/apis/table.entity.azure.data-tables.test.ts:1310]: Test cases added

File: tests/table/apis/table.entity.rest.test.ts (hunk 3 above)
Comment (edwin-huber) [reply tests/table/apis/table.entity.rest.test.ts:316]: The original test name was left over from when we had not implemented the functionality.
I shall improve the test.

File: tests/table/dotnet/AzuriteTableTest/Program.cs (hunk 5 above)
Comment (edwin-huber) [reply tests/table/dotnet/AzuriteTableTest/Program.cs:20]: These are not yet Tests, and will be converted to NUnit tests soon.

File: src/table/batch/TableBatchOrchestrator.ts (hunk 1 above)
Comment (edwin-huber) [reply src/table/batch/TableBatchOrchestrator.ts:662]: I'll add tests to cover Azure Storage SDK as Data-Tables is done

**Comment by blueww** [comment]
@edwin-huber 

I see you have resolved the comments and added test cases.
Feel free to let us know if the PR is ready for the 2nd round review and merge. And we will review it again after get your notice.

**Review by XiaoningLiu — APPROVED** [review]

**Review by blueww — APPROVED** [review]

**Pull request merged**